client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -0,0 +1,477 @@
+					&BlockId::Hash(notification.hash),
+					Some(&BlockId::Hash(*notification.header.parent_hash())),
+				);
+
+				let _ = subscriptions.pin_block(&sub_id_import, notification.hash);
+
+				// Note: `Block::Hash` will serialize to hexadecimal encoded string.
+				let new_block = FollowEvent::NewBlock(NewBlock {
+					block_hash: notification.hash,
+					parent_block_hash: *notification.header.parent_hash(),
+					new_runtime,
+					runtime_updates,
+				});
+
+				if !notification.is_new_best {
+					return stream::iter(vec![new_block])
+				}
+
+				// If this is the new best block, then we need to generate two events.
+				let best_block = FollowEvent::BestBlockChanged(BestBlockChanged {
+					best_block_hash: notification.hash,
+				});
+				stream::iter(vec![new_block, best_block])
+			})
+			.flatten();
+
+		let subscriptions = self.subscriptions.clone();
+		let sub_id_finalized = sub_id.clone();
+
+		let stream_finalized =
+			self.client.finality_notification_stream().map(move |notification| {
+				// We might not receive all new blocks reports, also pin the block here.
+				let _ = subscriptions.pin_block(&sub_id_finalized, notification.hash);
+
+				// The tree route contains the exclusive path from the latest finalized block
+				// to the block reported by the notification. Ensure the finalized block is
+				// properly reported to that path.
+				let mut finalized_block_hashes =
+					notification.tree_route.iter().cloned().collect::<Vec<_>>();
+				finalized_block_hashes.push(notification.hash);
+
+				FollowEvent::Finalized(Finalized {
+					finalized_block_hashes,
+					pruned_block_hashes: notification.stale_heads.iter().cloned().collect(),
+				})
+			});
+
+		let merged = tokio_stream::StreamExt::merge(stream_import, stream_finalized);
+
+		// The initialized event is the first one sent.
+		let finalized_block_hash = self.client.info().finalized_hash;
+		let finalized_block_runtime = generate_runtime_event(
+			&self.client,
+			runtime_updates,
+			&BlockId::Hash(finalized_block_hash),
+			None,
+		);
+
+		let _ = self.subscriptions.pin_block(&sub_id, finalized_block_hash);
+
+		let stream = stream::once(async move {
+			FollowEvent::Initialized(Initialized {
+				finalized_block_hash,
+				finalized_block_runtime,
+				runtime_updates,
+			})
+		})
+		.chain(merged);
+
+		let subscriptions = self.subscriptions.clone();
+		let fut = async move {
+			if let SubscriptionClosed::Failed(_) = sink.pipe_from_stream(stream.boxed()).await {
+				// The subscription failed to pipe from the stream.
+				let _ = sink.send(&FollowEvent::<Block::Hash>::Stop);
+			}
+
+			// The client disconnected or called the unsubscribe method.
+			subscriptions.remove_subscription(&sub_id);
+		};
+
+		self.executor.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
+		Ok(())
+	}
+
+	fn chain_head_unstable_body(
+		&self,
+		mut sink: SubscriptionSink,
+		follow_subscription: String,
+		hash: Block::Hash,
+		_network_config: Option<()>,
```

Comment:
I don't think this should be `()`.
The config should be ignored by the full node, but it seems to me that `()` would treat any object sent by the client as invalid?

Reply:
Yep, the RPC will reply with invalid parameters instead of ignoring the networkConfig. Have added a new type for this, thanks!